File: frontend/src/api_wrapper/auth/refresh_accesstoken.ts
import { makeRequest, type RequestResult } from '../common/request';
import { setCookie } from 'typescript-cookie';

interface RefreshAccessTokenPayload {
    refresh_token: string;
}

interface RefreshAccessTokenReturnPayload {
    access_token: string;
}

const REFRESH_ACCESSTOKEN_ROUTE = 'auth/refresh_token';

export async function refreshAccessToken(
    payload: RefreshAccessTokenPayload
): Promise<RequestResult<RefreshAccessTokenReturnPayload>> {
    const result = await makeRequest<RefreshAccessTokenPayload, RefreshAccessTokenReturnPayload>(
        payload,
        REFRESH_ACCESSTOKEN_ROUTE,
        false
    );

    if (result.success && result.value) {
        setCookie('accesstoken', result.value.access_token);
    }

    return result;
}

File: server/src/services/auth/refresh_accesstoken.rs
use crate::errors::Errors;
use crate::jwt::claims::TokenType;
use crate::jwt::sign::generate_accesstoken;
use crate::jwt::verify::verify_token;
use crate::services::auth::payloads::{RefreshAccessTokenPayload, RefreshAccessTokenReturnPayload};
use actix_web::{web::Json, Responder};

pub async fn refresh_accesstoken(
    payload: Json<RefreshAccessTokenPayload>,
) -> Result<impl Responder, Errors> {
    let refresh_token = payload.refresh_token.clone();
    let token_payload = verify_token(&refresh_token)
        .map_err(|_| Errors::BadRequest("Invalid token".to_string()))?;

    match token_payload.token_type {
        TokenType::RefreshToken => (),
        _ => return Err(Errors::BadRequest("Invalid token".to_string())),
    }

    let new_accesstoken = generate_accesstoken(token_payload.id.parse().unwrap())
        .map_err(|_| Errors::InternalServerError)?;

    Ok(Json(RefreshAccessTokenReturnPayload {
        access_token: new_accesstoken,
    }))
}
